cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll

```diff
@@ -713,11 +713,23 @@ private Node getAPriorDefinition(SsaDefOrUse defOrUse) {
 /** Holds if there is def-use or use-use flow from `nodeFrom` to `nodeTo`. */
 predicate ssaFlow(Node nodeFrom, Node nodeTo) {
   exists(Node nFrom, boolean uncertain, SsaDefOrUse defOrUse |
-    ssaFlowImpl(defOrUse, nFrom, nodeTo, uncertain) and
+    ssaFlowImpl(defOrUse, nFrom, nodeTo, uncertain) and nodeFrom != nodeTo
+  |
     if uncertain = true then nodeFrom = [nFrom, getAPriorDefinition(defOrUse)] else nodeFrom = nFrom
   )
 }
 
+predicate postUpdateFlow(PostUpdateNode pun, Node nodeTo) {
+  exists(Node preUpdate, Node nFrom, boolean uncertain, SsaDefOrUse defOrUse |
+    preUpdate = pun.getPreUpdateNode() and
+    ssaFlowImpl(defOrUse, nFrom, nodeTo, uncertain)
+  |
+    if uncertain = true
+    then preUpdate = [nFrom, getAPriorDefinition(defOrUse)]
+    else preUpdate = nFrom
+  )
+}
```

Comment:
Can you explain what the change is here / why `nodeFrom != nodeTo` above is insufficient on its own?

Otherwise LGTM.

Reply:
Sure. This was also my initial idea, but then `@hvitved` pointed out that for flow out of post-update nodes specifically we need to allow SSA identity-steps. To see why, consider this example:
```cpp
void setTaint(x) {
  sink(x.field);
  x.field = taint;
}

while (b) {
  setTaint(y);
}
```
the flow after leaving `setTaint` (i.e., from `[post] y`) to `y` is implemented as an SSA step from `[post] y`'s pre-update node (which is exactly `y`) to the next use of `y`, but the next use of `y` is `y` itself! So in order to have flow from `[post] y` to `y` the `ssaFlow` predicate must include steps from `y` to `y` itself in order to support flow out of post-update nodes.

So the final creates a new `postUpdateFlow` predicate whose implementation is equivalent to `ssaFlow`, but allow identity steps.

Does that make sense?

Reply:
That makes sense, yep.  Thanks for the explanation!